source/puppet/4.4/reference/static_catalogs.md

```diff
@@ -0,0 +1,133 @@
+
+When applying a file resource from a static catalog, an agent first checks the catalog for that file's metadata. If it finds some, Puppet uses the metadata to call the [`static_file_content`][] API endpoint on the Puppet Server and retrieve the file's contents, also called the `code_content`. If the catalog doesn't contain metadata for the resource, Puppet does what it's always done: request the file resource's metadata from the master, compare it to the local file if it exists, and request the resource's file from the master in its current state if the local file doesn't exist or differs from the master's version.
+
+### Configuring `code_id` and the `static_file_content` endpoint
+
+When requesting the file's content via the static catalog's metadata, the Puppet agent passes the file's path, the catalog's `code_id`, and the requested environment to Puppet Server's new [`static_file_content`][] API endpoint, and the endpoint returns the appropriate version of the file's contents as the `code_content`.
+
+If static catalogs are enabled but none of the relevant Puppet Server settings are configured, the `code_id` parameter defaults to a null value and the agent uses the [`file_content`][] API endpoint, which always returns the latest content. To populate the `code_id` with a more useful identifier and have the agent use the  `static_file_content` endpoint to retrieve a specific version of the file's content, you must specify scripts or commands that provide Puppet with the appropriate results.
+
+Puppet Server locates these commands via the `code-id-command` and `code-content-command` settings in Puppet Server's [`puppetserver.conf`][] file. Puppet Server runs the `code-id-command` each time it compiles a static catalog, and it runs the `code-content-command` each time an agent requests file contents from the `static_file_content` endpoint.
+
+> **Note: The Puppet Server process must be able to execute these scripts. Puppet Server also validates their output and checks their exit codes. Environment names can contain only alphanumeric characters and underscores (`_`). The `code_id` can  contain only alphanumeric characters and dashes (`-`), underscores (`_`), semicolons (`;`), and colons (`:`). If either command returns a non-zero exit code, Puppet Server logs an error and returns the error message and a 500 response code to the API request.
+
+Puppet Server validates the standard output of each of these scripts, and if the output's acceptable, it adds the results to the catalog as their respective parameters' values. This lets you use any versioning or synchronization tools you want, as long as you can write scripts that produce a hash for the `code_id` and code content using the catalog's `code_id` and file's environment.
```

Comment:
This still uses the word `hash`.  Sorry, not sure why I pointed this out last time around, but I'm still thinking it's better to avoid that word here.